File: scripts/openapi-e2e/go/main.go
package main

import (
	"bytes"
	"context"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	openapi "github.com/GIT_USER_ID/GIT_REPO_ID"
)

var checks = []string{
	"generated-model-create",
	"bearer-auth-inspect",
	"typed-command-execution",
	"lossless-base64-output",
	"typed-file-roundtrip",
	"delete-and-stale-handle-rejection",
	"credential-redaction",
}

type evidence struct {
	SchemaVersion int      `json:"schema_version"`
	Language      string   `json:"language"`
	SandboxID     string   `json:"sandbox_id"`
	Checks        []string `json:"checks"`
}

func require(condition bool, message string) {
	if !condition {
		panic(message)
	}
}

func main() {
	apiURL := os.Getenv("FERROBOX_API_URL")
	auditPath := os.Getenv("FERROBOX_AUDIT_LOG")
	evidencePath := os.Getenv("FERROBOX_OPENAPI_SDK_EVIDENCE")
	require(apiURL != "" && auditPath != "" && evidencePath != "", "required environment is absent")

	configuration := openapi.NewConfiguration()
	configuration.Servers[0].URL = apiURL
	client := openapi.NewAPIClient(configuration)

	network := openapi.NewNetworkRequest()
	network.SetInternetAccess(false)
	createRequest := openapi.NewCreateSandboxRequest("python")
	createRequest.SetCpuCount(1)
	createRequest.SetMemoryMb(512)
	createRequest.SetTimeoutSeconds(120)
	createRequest.SetNetwork(*network)
	created, response, err := client.SandboxesAPI.CreateSandbox(context.Background()).
		CreateSandboxRequest(*createRequest).
		Execute()
	require(err == nil, fmt.Sprintf("create failed: %v", err))
	require(response.StatusCode == 201, fmt.Sprintf("create returned %d", response.StatusCode))
	require(created.GetState() == openapi.RUNNING, "created sandbox is not running")

	sandboxID := created.GetSandboxId()
	token := created.GetToken()
	require(sandboxID != "" && token != "", "create identity is absent")
	ownedContext := context.WithValue(context.Background(), openapi.ContextAccessToken, token)
	deleted := false
	defer func() {
		if !deleted {
			_, _ = client.SandboxesAPI.DeleteSandbox(ownedContext, sandboxID).Execute()
		}
	}()

	inspected, response, err := client.SandboxesAPI.GetSandbox(ownedContext, sandboxID).Execute()
	require(err == nil, fmt.Sprintf("inspect failed: %v", err))
	require(response.StatusCode == 200, fmt.Sprintf("inspect returned %d", response.StatusCode))
	require(inspected.GetSandboxId() == sandboxID, "inspect returned another sandbox")
	require(inspected.GetState() == openapi.RUNNING, "inspected sandbox is not running")

	executeRequest := openapi.NewExecuteCommandRequest([]string{"python3", "-c", "print(40 + 2)"})
	executeRequest.SetCwd("/home/sandbox")
	executeRequest.SetEnvironment(map[string]string{})
	executeRequest.SetTimeoutSeconds(30)
	executeRequest.SetMaxOutputBytes(1048576)
	executed, response, err := client.CommandsAPI.ExecuteCommand(ownedContext, sandboxID).
		ExecuteCommandRequest(*executeRequest).
		Execute()
	require(err == nil, fmt.Sprintf("execute failed: %v", err))
	require(response.StatusCode == 200, fmt.Sprintf("execute returned %d", response.StatusCode))
	require(executed.GetStdout() == "42\n", "typed stdout mismatch")
	decodedStdout, err := base64.StdEncoding.DecodeString(executed.GetStdoutBase64())
	require(err == nil && string(decodedStdout) == "42\n", "base64 stdout mismatch")

	payload := []byte("generated-openapi-client\n")
	writeRequest := openapi.NewWriteFileRequest(
		"/home/sandbox/openapi.txt",
		base64.StdEncoding.EncodeToString(payload),
	)
	writeRequest.SetOverwrite(false)
	written, response, err := client.FilesAPI.WriteFile(ownedContext, sandboxID).
		WriteFileRequest(*writeRequest).
		Execute()
	require(err == nil, fmt.Sprintf("write failed: %v", err))
	require(response.StatusCode == 200, fmt.Sprintf("write returned %d", response.StatusCode))
	require(written.GetBytesWritten() == int64(len(payload)), "written byte count mismatch")

	read, response, err := client.FilesAPI.ReadFile(ownedContext, sandboxID).
		Path("/home/sandbox/openapi.txt").
		Offset(0).
		MaxBytes(1048576).
		Execute()
	require(err == nil, fmt.Sprintf("read failed: %v", err))
	require(response.StatusCode == 200, fmt.Sprintf("read returned %d", response.StatusCode))
	content, err := base64.StdEncoding.DecodeString(read.GetContentBase64())
	require(err == nil && bytes.Equal(content, payload), "file content mismatch")
	require(read.GetEof(), "file read did not reach EOF")

	response, err = client.SandboxesAPI.DeleteSandbox(ownedContext, sandboxID).Execute()
	require(err == nil, fmt.Sprintf("delete failed: %v", err))
	require(response.StatusCode == 204, fmt.Sprintf("delete returned %d", response.StatusCode))
	deleted = true
	_, response, err = client.SandboxesAPI.GetSandbox(ownedContext, sandboxID).Execute()
	require(err != nil && response != nil && response.StatusCode == 404, "deleted sandbox remained addressable")

	audit, err := os.ReadFile(auditPath)
	require(err == nil, fmt.Sprintf("read audit: %v", err))
	require(!strings.Contains(string(audit), token), "bearer credential reached audit log")
	require(strings.Contains(string(audit), `"operation":"delete"`), "delete audit is absent")

	result := evidence{SchemaVersion: 1, Language: "go", SandboxID: sandboxID, Checks: checks}
	encoded, err := json.MarshalIndent(result, "", "  ")
	require(err == nil, fmt.Sprintf("encode evidence: %v", err))
	encoded = append(encoded, '\n')
	require(os.WriteFile(evidencePath, encoded, 0o644) == nil, "write evidence failed")
	fmt.Print(string(encoded))
}

File: scripts/openapi-e2e/rust/main.rs
use std::{env, error::Error, fs};

use base64::{engine::general_purpose::STANDARD as BASE64, Engine as _};
use ferrobox_client::{
    apis::{
        commands_api, configuration::Configuration, files_api, sandboxes_api, Error as ApiError,
    },
    models,
};
use serde_json::json;

const CHECKS: [&str; 7] = [
    "generated-model-create",
    "bearer-auth-inspect",
    "typed-command-execution",
    "lossless-base64-output",
    "typed-file-roundtrip",
    "delete-and-stale-handle-rejection",
    "credential-redaction",
];

fn required_environment(name: &str) -> Result<String, Box<dyn Error>> {
    env::var(name).map_err(|_| format!("{name} is required").into())
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let api_url = required_environment("FERROBOX_API_URL")?;
    let audit_path = required_environment("FERROBOX_AUDIT_LOG")?;
    let evidence_path = required_environment("FERROBOX_OPENAPI_SDK_EVIDENCE")?;

    let mut public_configuration = Configuration::new();
    public_configuration.base_path = api_url.clone();
    let mut create_request = models::CreateSandboxRequest::new("python".to_owned());
    create_request.cpu_count = Some(1);
    create_request.memory_mb = Some(512);
    create_request.timeout_seconds = Some(120);
    let mut network = models::NetworkRequest::new();
    network.internet_access = Some(false);
    create_request.network = Some(Box::new(network));
    let created = sandboxes_api::create_sandbox(&public_configuration, create_request).await?;
    if created.state != models::SandboxState::Running || created.token.is_empty() {
        return Err("created sandbox identity is invalid".into());
    }

    let sandbox_id = created.sandbox_id.to_string();
    let token = created.token;
    let mut configuration = Configuration::new();
    configuration.base_path = api_url;
    configuration.bearer_access_token = Some(token.clone());

    let inspected = sandboxes_api::get_sandbox(&configuration, &sandbox_id).await?;
    if inspected.sandbox_id != created.sandbox_id
        || inspected.state != models::SandboxState::Running
    {
        return Err("inspect returned invalid sandbox state".into());
    }

    let mut execute_request = models::ExecuteCommandRequest::new(vec![
        "python3".to_owned(),
        "-c".to_owned(),
        "print(40 + 2)".to_owned(),
    ]);
    execute_request.cwd = Some("/home/sandbox".to_owned());
    execute_request.environment = Some(Default::default());
    execute_request.timeout_seconds = Some(30);
    execute_request.max_output_bytes = Some(1_048_576);
    let executed =
        commands_api::execute_command(&configuration, &sandbox_id, execute_request).await?;
    if executed.stdout != "42\n" || BASE64.decode(&executed.stdout_base64)? != b"42\n" {
        return Err("command output mismatch".into());
    }

    let payload = b"generated-openapi-client\n";
    let mut write_request = models::WriteFileRequest::new(
        "/home/sandbox/openapi.txt".to_owned(),
        BASE64.encode(payload),
    );
    write_request.overwrite = Some(false);
    let written = files_api::write_file(&configuration, &sandbox_id, write_request).await?;
    if written.bytes_written != payload.len() as i64 {
        return Err("written byte count mismatch".into());
    }

    let read = files_api::read_file(
        &configuration,
        &sandbox_id,
        "/home/sandbox/openapi.txt",
        Some(0),
        Some(1_048_576),
    )
    .await?;
    if BASE64.decode(&read.content_base64)? != payload || !read.eof {
        return Err("file roundtrip mismatch".into());
    }

    sandboxes_api::delete_sandbox(&configuration, &sandbox_id).await?;
    let stale = sandboxes_api::get_sandbox(&configuration, &sandbox_id).await;
    match stale {
        Err(ApiError::ResponseError(response)) if response.status.as_u16() == 404 => {}
        _ => return Err("deleted sandbox remained addressable".into()),
    }

    let audit = fs::read_to_string(audit_path)?;
    if audit.contains(&token) || !audit.contains("\"operation\":\"delete\"") {
        return Err("audit credential-redaction check failed".into());
    }

    let evidence = json!({
        "schema_version": 1,
        "language": "rust",
        "sandbox_id": sandbox_id,
        "checks": CHECKS,
    });
    let serialized = serde_json::to_string_pretty(&evidence)? + "\n";
    fs::write(evidence_path, &serialized)?;
    print!("{serialized}");
    Ok(())
}
